### PR TITLE
drivers: can: CANBUS driver for native_posix board

### PR DESCRIPTION
**Pull request by pllee4-weston**
- Rebased jukkar's work for canbus for native_posix board to latest main branch

https://github.com/jukkar/zephyr/pull/21
https://github.com/zephyrproject-rtos/zephyr/pull/14196

@jukkar it would be great if you can help to review as well

### CODE REVIEW

**Comment by rlubos** [comment]
@cfriedt I'm not a CAN maintainer, I think it should rather be assigned to @alexanderwachter.

File: drivers/can/CMakeLists.txt
Comment (gmarull) [inline drivers/can/CMakeLists.txt:16]: Just a quick comment: socket CAN is only available on Linux, so I think a name other than "native_posix" should be used. The build should only be available on Linux hosts actually (maybe `CMAKE_HOST_SYSTEM_NAME` can be useful here?)

File: drivers/can/Kconfig.native_posix
Comment (gmarull) [inline drivers/can/Kconfig.native_posix:34]: DeviceTree can likely be used for that

**Comment by alexanderwachter** [comment]
SocketCAN was written mainly maintained by @jukkar. I'm not very familiar with SocketCAN, but I can comment on the API later this week.

**Comment by pfalcon** [comment]
@pllee4-weston: This now has a git conflict, please rebase.

**Comment by pfalcon** [comment]
@alexanderwachter on Jul 26:

>  I'm not very familiar with SocketCAN, but I can comment on the API later this week.

It still would be nice to get a review from you, as you're the CAN guy ;-).

File: drivers/can/can_native_posix.c
Comment (alexanderwachter) [inline drivers/can/can_native_posix.c:16]: These can go to the registration in line 20

File: drivers/can/can_native_posix.c
Comment (alexanderwachter) [inline drivers/can/can_native_posix.c:88]: Why is there a sleep? Do you need to poll for the interface up?

File: drivers/can/can_native_posix.c
Comment (alexanderwachter) [inline drivers/can/can_native_posix.c:100]: I would rather return an error in case someone pass a callback.
Since callbacks are not supported, this driver may not work with other CAN subsys like ISO-TP or canopen.

File: drivers/can/can_native_posix.c
Comment (alexanderwachter) [inline drivers/can/can_native_posix.c:118]: How do you want to receive any frame if this function is empty?
Seems that you only want to support socketCAN. If so, please remove the z_can API

File: drivers/can/can_native_posix.c
Comment (alexanderwachter) [inline drivers/can/can_native_posix.c:173]: Return an error?

File: drivers/can/can_native_posix_adapt.c
Comment (alexanderwachter) [inline drivers/can/can_native_posix_adapt.c:40]: Same as above, move the name and level as arguments to this macro

File: drivers/can/can_native_posix_adapt.c
Comment (alexanderwachter) [inline drivers/can/can_native_posix_adapt.c:67]:
```suggestion
		close(fd);
		return -errno;
```

**Review by alexanderwachter — CHANGES_REQUESTED** [review]

**Comment by alexanderwachter** [comment]
Since this PR only supports socketCAN, why don't you use a socket with PF_CAN directly instead?

**Pull request closed without merging**
